### PR TITLE
Move away from cohttp to libcurl

### PR DESCRIPTION
**Pull request by eXenon**

### CODE REVIEW

File: cs_api_io/cs_api_io.ml [hunk 1]
```diff
@@ -1,79 +1,53 @@
...
-    let message =
-      Printf.sprintf "%s\n%s" (Cohttp.Code.string_of_status status) body_str
-    in
-    Lwt_result.fail message
+let get_response {Response.code; response} =
+  if code < 300 then
+    Lwt_result.return !response
   else
-    Lwt_result.return body_str
+    Lwt_result.fail !response
```
Comment (bbc2) [inline cs_api_io/cs_api_io.ml:53]: What about keeping the same sort of message as before when there's a failure, with status code and body?

File: cs_api_io/cs_api_io.ml [hunk 2]
```diff
@@ -1,79 +1,53 @@
...
-    | Multipart parts -> handle_multipart parts )
-    >>= fun (extra_headers, body) ->
-    let headers = Cohttp.Header.add_list headers extra_headers in
-    let body = Cohttp_lwt.Body.of_stream body in
-    Lwt_result.ok (Cohttp_lwt_unix.Client.post ~ctx ~body ~headers url)
+  try
+    Curl.perform curl;
+    Lwt_result.return {Response.code = Curl.get_responsecode curl; response}
+  with Curl.CurlException (_, _, s) ->
+    Lwt_result.fail ("HTTP exception : " ^ s)
```
Comment (bbc2) [inline cs_api_io/cs_api_io.ml:45]: - What about "HTTP error"?   I think would make more sense to the user than "exception" in the context of HTTP.
- As you probably know now, unlike in French, there's no space before a colon or semi-colon character in English.

File: cs_api_io/cs_api_io.ml [hunk 3]
```diff
@@ -1,79 +1,53 @@
-let get_ctx ~verify =
-  let open Lwt.Infix in
-  Conduit_lwt_unix.init ~verify () >|= fun ctx ->
-  Cohttp_lwt_unix.Client.custom_ctx ~ctx ()
+module Response = struct
+  type t =
+    { code : int
+    ; response : string ref }
```
Comment (bbc2) [inline cs_api_io/cs_api_io.ml:4]: - As far as I can tell, there's nothing in the code that requires this to be a `ref`.  If that's indeed the case, what about making it a simple string?
- What about renaming this record field to `body`?

**Review by bbc2 — CHANGES_REQUESTED** [review]
- The libcurl stuff is nicely contained.  Good job with that!
- Have you had the time to test HTTP proxying?  Does it work out of the box?  If yes, I would suggest adding that to the changelog, the commit message (someone reading it later would probably not understand why we moved to libcurl) and perhaps the readme (at least to say that we can do the same kind of proxying as with curl, if that's the case, or even what environment variable to set to make it work).

File: cs_api_io/cs_api_io.ml (hunk 2 above)
Comment (eXenon) [reply cs_api_io/cs_api_io.ml:45]: I'll never get used to the colons :(... Fixed.

File: cs_api_io/cs_api_io.ml (hunk 3 above)
Comment (eXenon) [reply cs_api_io/cs_api_io.ml:4]: Yes, good catch, that is a left-over of an older version. Changed it to match your suggestions.

**Comment by eXenon** [comment]
- Thanks :) !
- Yes, I was able to test this properly. It works without any problems, the proxy correctly routes the request as soon as the environment variable is set.

I have added a section in the README for users wanting to configure the API client for use with proxies. I was thinking that maybe we should also add a note about the dependency on libcurl? As for the motivation for this swtich, I have edited the commit message to include it.

**Comment by bbc2** [comment]
>  I was thinking that maybe we should also add a note about the dependency on libcurl?

Good idea.

File: cs_api_io/cs_api_io.ml [hunk 4]
```diff
@@ -1,79 +1,65 @@
...
     let message =
-      Printf.sprintf "%s\n%s" (Cohttp.Code.string_of_status status) body_str
+      match s with
+      | "CURLE_URL_MALFORMAT" -> "Malformed URL"
+      | "CURLE_COULDNT_CONNECT" -> "Could not connect to host or proxy"
+      | "CURLE_OPERATION_TIMEDOUT" -> "Request timed out"
+      | "CURLE_SEND_ERROR" -> "Unable to send data to the network"
+      | "CURLE_RECV_ERROR" -> "Unable to receive data from the network"
+      | "CURLE_SSL_INVALIDCERTSTATUS" -> "SSL certificate could not be verified"
+      | _ -> "Unknown error:" ^ s
```
Comment (bbc2) [inline cs_api_io/cs_api_io.ml:54]: Nitpick: missing space after the `:`

**Review by bbc2 — CHANGES_REQUESTED** [review]
I think you're still planning to update the changelog about the new dependency and the new proxying capability.  Otherwise it's pretty much ready to merge.

**Review by bbc2 — APPROVED** [review]

**Pull request merged**